R/linter_tags.R

```diff
@@ -0,0 +1,186 @@
+#' Get Linter metadata from a package
+#'
+#' Obtain a tagged list of all Linters available in a package.
+#'
+#' @param packages A character vector of packages to search for linters.
+#'
+#' @section Package Authors:
+#'
+#' To implement \code{available_linters()} for your package, include a file \code{inst/lintr/linters.csv} in your
+#' package.
+#' The CSV file must contain two columns, 'linter' and 'tags', and be UTF-8 encoded.
+#' Each row describes a linter by
+#'
+#' \enumerate{
+#' \item{its function name (e.g. \code{"assignment_linter"}) in the column 'linter'.}
+#' \item{space separated tags associated with the linter (e.g. \code{"style consistency default"}) in the column
+#'   'tags'.}
+#' }
+#'
+#' Tags should be snake_case.
+#'
+#' @return
+#' A data frame with columns 'linter', 'package' and 'tags':
+#'
+#' \describe{
+#' \item{linter}{A character column naming the function associated to the linter.}
+#' \item{package}{A characetr column containing the name of the package providing the linter.}
+#' \item{tags}{A list column containing tags associated to the linter.}
+#' }
+#'
+#' @examples
+#' lintr_linters <- available_linters()
+#'
+#' # If the package doesn't exist or isn't installed, an empty data frame will be returned
+#' available_linters("does-not-exist")
+#'
+#' lintr_linters2 <- available_linters(c("lintr", "does-not-exist"))
+#' identical(lintr_linters, lintr_linters2)
+#' @export
+available_linters <- function(packages = "lintr") {
+  if (!is.character(packages)) {
+    stop("`packages` must be a character vector.")
+  }
+
+  # Handle multiple packages
+  if (length(packages) > 1L) {
+    return(do.call(rbind, lapply(packages, available_linters)))
+  }
+
+  csv_file <- system.file("lintr", "linters.csv", package = packages)
+
+  # Make sure we always return a valid data frame
+  # data.frame(...) and as.data.frame(...) don't handle zero-row list-columns properly, so we need to manually create
+  # the structure.
+  empty_linters <- structure(list(
+    linter = character(),
+    package = character(),
+    tags = list()
+  ), row.names = integer(0), class = "data.frame")
+
+  if (!file.exists(csv_file)) {
+    return(empty_linters)
+  }
+  available <- utils::read.csv(csv_file, encoding = "UTF-8", as.is = TRUE)
+
+  # Check that the csv file contains two character columns, named 'linter' and 'tags'.
+  # Otherwise, fallback to an empty data frame.
+  if (!ncol(available) == 2L) {
```

Comment:
I think we can be more flexible here, maybe downstream will eventually add their own columns. We can just search for `linter`/`tags` in the columns and return those columns, only error if those two aren't present.

Reply:
changed to requiring those two columns